torch/csrc/jit/passes/break_continue_transform.cpp

```diff
@@ -0,0 +1,326 @@
+    auto loop_status = WONT;
+    for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+      Node* node = *it;
+      it++;
+      switch (node->kind()) {
+        case prim::Function: {
+          handleTransforms(node->blocks().at(0));
+        } break;
+        case prim::ContinueStmt:
+        case prim::BreakStmt: {
+          if (node->kind() != transformKind()) {
+            continue;
+          }
+          WithInsertPoint b(block);
+          node->destroy();
+          loop_status = WILL;
+        } break;
+        case prim::If: {
+          loop_status = handleIf(node);
+        } break;
+        case prim::Loop: {
+          handleLoop(node);
+        } break;
+      }
+      if (loop_status == WILL) {
+        deleteAfterBreakNodes(block, it);
+        break;
+      } else if (loop_status == MIGHT) {
+        if (it != block->nodes().end()) {
+          loop_status = guardBlockNodes(block, it);
+        }
+        break;
+      }
+    }
+
+    {
+      // MIGHT value must be an output of an if, so we do not need to set it
+      WithInsertPoint insert(block);
+      if (loop_status == WILL) {
+        graph_->insertNode(graph_->createStore(getVarname(), true_val_));
+      } else if (loop_status == WONT) {
+        graph_->insertNode(graph_->createStore(getVarname(), false_val_));
+      }
+    }
+
+    return loop_status;
+  }
+
+  void run() {
+    handleTransforms(graph_->block());
+  }
+
+  Transform transform_;
+  Value* true_val_ = nullptr;
+  Value* false_val_ = nullptr;
+
+  std::shared_ptr<Graph> graph_;
+};
+
+void moveBlockBeforeNode(Node* before_node, Block* block) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    auto block_node = *it++;
+    block_node->moveBefore(before_node);
+  }
+}
+
+// The loop node is initially emitted as:
+// Loop(max_trip_count)
+//    block0(loop_counter) {
+//      <body>
+//    }
+//    block1 {
+//      <loop condition>
+//      -> (condition)
+//    }
+// Here, we inline the loop condition into:
+// Loop(max_trip_count, start_condition)
+//    block0(loop_counter) {
+//      <body>
+//    }
+//    block1 {
+//      <loop condition>
+//      -> (condition)
+//    }
+
+void inlineLoopStartCondition(Node* n) {
+  auto pre_header = n->blocks().at(1);
+  auto header_block = n->addBlock();
+  header_block->cloneFrom(pre_header, [](Value* v) { return v; });
+  moveBlockBeforeNode(n, header_block);
+  n->addInput(header_block->outputs().at(0));
+  n->eraseBlock(2);
+}
+
+void inlineLoopStartCondition(Block* block) {
+  for (Node* n : block->nodes()) {
+    switch (n->kind()) {
+      case prim::If:
+      case prim::Function: {
+        for (auto b : n->blocks()) {
+          inlineLoopStartCondition(b);
+        }
+      } break;
+      case prim::Loop: {
+        inlineLoopStartCondition(n->blocks().at(0));
+        inlineLoopStartCondition(n);
+      } break;
+    }
+  }
+}
+
+// First we inline the loop input condition.
```

Comment:
This organization is confusing. I would not expect a pass that requires break/continue loops to also mess with the format of the loop conditions.